lib/SILOptimizer/Mandatory/MandatoryCombine.cpp

```diff
@@ -284,6 +281,86 @@ SILInstruction *MandatoryCombiner::visitApplyInst(ApplyInst *instruction) {
   return nullptr;
 }
 
+template <class InstT> static FunctionRefInst *getRemovableRef(InstT *i) {
+  // If the only use of the function_ref is us, then remove it.
+  auto funcRef = dyn_cast<FunctionRefInst>(i->getCallee());
+  if (funcRef &&
+      (funcRef->use_empty() ||
+       (funcRef->getSingleUse() && funcRef->getSingleUse()->getUser() == i))) {
+    return funcRef;
+  }
+  return nullptr;
+}
+
+SILInstruction *MandatoryCombiner::visitLoadInst(LoadInst *i) {
+  auto val = cleanupLoadedCalleeValue(i, instModCallbacks.deleteInst);
+  if (!val)
+    return nullptr;
+
+  val = stripCopiesAndBorrows(val);
+  auto src = val;
+
+  // Try to get from a convert_function/convert_escape_to_noescape to a
+  // partial_apply/thin_to_thick to a convert_function.
+  if (auto *cfi = dyn_cast<ConvertFunctionInst>(val))
+    src = stripCopiesAndBorrows(cfi->getOperand());
+
+  if (auto *cvt = dyn_cast<ConvertEscapeToNoEscapeInst>(val))
+    src = stripCopiesAndBorrows(cvt->getOperand());
+
+  if (auto *pa = dyn_cast<PartialApplyInst>(src)) {
+    if (tryDeleteDeadClosure(pa, instModCallbacks)) {
+      val = pa->getCallee();
+      if (auto *ref = getRemovableRef(pa)) {
+        instModCallbacks.deleteInst(ref);
+        return nullptr;
+      }
+    }
+  } else if (auto *tttf = dyn_cast<ThinToThickFunctionInst>(src)) {
+    if (tryDeleteDeadClosure(tttf, instModCallbacks)) {
```

Comment:
Most of the optimizations here should not be on LoadInst. You aren't removing load inst anywhere! Instead, all of these should be on the actual instructions you are removing (i.e. ThinToThick, Partial, Etc).